File: agent/commands-generator-api/src/main/java/org/matilda/commands/CommandRepository.java
package org.matilda.commands;

import javax.inject.Inject;
import javax.inject.Singleton;
import java.util.HashMap;
import java.util.Map;

@Singleton
public class CommandRepository {
    private static final int DEFAULT_PROVIDER_ID = 0;
    private final Map<Integer, CommandRegistry> mCommandsRegistries;

    @Inject
    CommandRegistryIdGenerator mCommandRegistryIdGenerator;

    @Inject
    public CommandRepository() {
        mCommandsRegistries = new HashMap<>();
    }

    public int addCommandRegistry(CommandRegistry commandRegistry) {
        int id = mCommandRegistryIdGenerator.generate();
        mCommandsRegistries.put(id, commandRegistry);
        return id;
    }

    public void setDefaultCommandRegistry(CommandRegistry commandRegistry) {
        mCommandsRegistries.put(DEFAULT_PROVIDER_ID, commandRegistry);
    }

    public CommandRegistry getCommandRegistry(int registryId) {
        if (!mCommandsRegistries.containsKey(registryId)) {
            throw new IllegalArgumentException(String.format("Command Provider %d not found", registryId));
        }
        return mCommandsRegistries.get(registryId);
    }

    public Command getCommand(int registryId, int commandType) {
        return getCommandRegistry(registryId).get(commandType);
    }
}

File: agent/library/src/commonMain/kotlin/org/matilda/commands/CommandRepository.kt
package org.matilda.commands

import me.tatarka.inject.annotations.Inject

@MatildaScope
@Inject
class CommandRepository(private val mCommandRegistryIdGenerator: CommandRegistryIdGenerator) : CommandRegistryManager {
    private val mCommandsRegistries = mutableMapOf<Int, CommandRegistry>()

    override fun addCommandRegistry(commandRegistry: CommandRegistry) = mCommandRegistryIdGenerator.generate().also {
        mCommandsRegistries[it] = commandRegistry
    }

    fun setDefaultCommandRegistry(commandRegistry: CommandRegistry) {
        mCommandsRegistries[DEFAULT_PROVIDER_ID] = commandRegistry
    }

    private fun getCommandRegistry(registryId: Int) = mCommandsRegistries.getOrElse(registryId) {
        throw IllegalArgumentException("Command Provider $registryId not found")
    }

    fun getCommand(registryId: Int, commandType: Int): Command {
        return getCommandRegistry(registryId)[commandType]
    }

    companion object {
        private const val DEFAULT_PROVIDER_ID = 0
    }
}
